src/common/config_parser.h

```diff
@@ -16,48 +13,74 @@
 
 namespace marian {
 
-enum struct ConfigMode {
-  training,
-  translating,
-  rescoring,
-};
-
-// try to determine the width of the terminal
-uint16_t guess_terminal_width(uint16_t max_width = 180);
+namespace cli {
+enum struct mode { training, translation, scoring };
+}  // namespace cli
 
+/**
+ * @brief Command-line options parser
+ *
+ * New options and aliases should be defined within `addOptions*` methods.
+ */
 class ConfigParser {
 public:
-  ConfigParser(int argc, char** argv, ConfigMode mode, bool validate = false)
-      : mode_(mode),
-        cmdline_options_("Allowed options", guess_terminal_width()) {
+  ConfigParser(int argc, char** argv, cli::mode mode, bool validate = false)
+      : mode_(mode) {
     parseOptions(argc, argv, validate);
   }
 
+  /**
+   * @brief Parse command-line options
+   *
+   * Options are parsed in the following order, later config options overwrite
+   * earlier:
+   *  * predefined default values
+   *  * options from the config files provided with --config, from left to right
+   *  * options from the model config file, e.g. model.npz.yml
+   *  * aliases expanded into options, e.g. --best-deep
+   *  * options provided as command-line arguments
+   *
+   * Parsed options are available from getConfig().
+   *
+   * @param argc
+   * @param argv
+   * @param validate Do or do not validate parsed options
+   */
   void parseOptions(int argc, char** argv, bool validate);
 
   YAML::Node getConfig() const;
   std::vector<DeviceId> getDevices();
 
+  // Check if the config has a boolean option for the given key and it is
```

Comment:
What if it has such option but it is not a boolean? I kind of lost the context where this is happening, but it seems that if it is next to parseOptions(), it would be clearer to rename this to hasBoolean()

Reply:
Fixed